Update Imperative-Shadow-DOM-Distribution-API.md

yuzhe-han opened the pull request:
Updated based on the TPAC F2F meeting in Sept, 2019. https://github.com/whatwg/html/issues/3534#issuecomment-537802687

yuzhe-han committed:
Update Imperative-Shadow-DOM-Distribution-API.md

Updated based on the TPAC F2F meeting in Sept, 2019. https://github.com/whatwg/html/issues/3534#issuecomment-537802687

JanMiksovsky commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
```diff
@@ -1,48 +1,70 @@
 
-- *manually-assigned-nodes* are different than [assigned nodes]. For example, users can pass a node which is not host's children. We don't throw an exception even if such a *invalid* node is passed, however, *invalid* node is never selected as [assigned nodes]; the engine recaluculate [assigned nodes] later, based on *manually-assigned-nodes*. Invalid nodes in *manually-assigned-nodes* are simply ...
+Thus, we don't allow mixing the declarative and imperative APIs in the same shadow tree. Web developers must explicitly opt-in
+to use an imperative API for each shadow tree. They do this with a new "slotting" parameter on attachShadow():
```
Most of this proposal (and the existing spec) talks about _assignment_ rather than "slotting". Recommend reflecting that by changing the option to `assignment` (or maybe `slotAssignment`, etc.):

```this.attachShadow({ assignment: 'manual' })```

That makes the parameter's connection to the `assign()`, `assignedNodes`, and `assignedElements` more direct.

yuzhe-han replied:
Makes sense.  I like the slotAssignment parameter because it's more descriptive of what type of assignment it is.

yuzhe-han replied:
Updated doc to  use 'slotAssignment' instead.

yuzhe-han committed:
Update Imperative-Shadow-DOM-Distribution-API.md 

Updated slotting parameter to slotAssignment so it's more aligned with other APIs,  assign(), assignedNodes().

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
```diff
@@ -1,48 +1,70 @@
 
-The straw-man proposal for Imperative Shadow DOM distribution API.
-The context is: [https://github.com/whatwg/html/issues/3534](https://github.com/whatwg/html/issues/3534)
+This is a strawman proposal for Imperative Shadow DOM distribution API.
```
```suggestion
This is a strawperson proposal for Imperative Shadow DOM distribution API.
```

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
```diff
@@ -1,48 +1,70 @@
-The straw-man proposal for Imperative Shadow DOM distribution API.
-The context is: [https://github.com/whatwg/html/issues/3534](https://github.com/whatwg/html/issues/3534)
+This is a strawman proposal for Imperative Shadow DOM distribution API.
+For more context, please see [the WhatWG issue](https://github.com/whatwg/html/issues/3534).
```
```suggestion
For more context, please see [the WHATWG HTML Standard issue](https://github.com/whatwg/html/issues/3534).
```

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
````diff
@@ -1,48 +1,70 @@
+One of the drawbacks of Shadow DOM v1, when compared to Shadow DOM v0, is that web developers have to specify slot= attribute for every shadow host's children (except for elements for the *default* slot).
+ 
+#### Case 1:  Slot attributes is required in markup.
 ```
````
```suggestion
```html
```

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
````diff
@@ -1,48 +1,70 @@
+Blink has a special logic for some built-in elements to control node-to-slot mapping.
+ 
+#### Case 2: Can’t slot based on condition. 
+``` 
````
Style issue: Use `html` after the three backticks, and fix all the curly quotes here to be normal quotes.

yuzhe-han replied:
Updated to use html and fixed all curly quotes to use normal quotes instead.

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
````diff
@@ -1,48 +1,70 @@
+   <div slot=”slot1”></div>
+</shadow-host>
+ ```
+The second issue is that component creators can’t change the slotting behavior based on condition. A component that wants to display a subset of slottable based on a property can’t easily be done via declarative slotting.
````
It'd be good to explain the connection to the above markup. Something like "So, there is no way to implement the `num-to-show` attribute in the above example.

It'd be even better if the example were more realistic, e.g. using an attribute like "tab" instead of "num-to-show" and a tag name like "tab-panel" instead of "shadow-host".

yuzhe-han replied:
Thanks for the suggested example. Updated my example to be a better case. Also added an example implementation to showcase the new API.

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
```diff
@@ -1,48 +1,70 @@
-The default is "we support only declarative API for the shadow tree".
+[assigned nodes]: https://dom.spec.whatwg.org/#slot-assigned-nodes
 
+An `HTMLSlotElement` gets a new API, called `assign()` (tentative name), and a new property, `manually-assigned-nodes`,
```
Properties in JavaScript are usually named in camel case, e.g. `slot.manuallyAssignedNodes`. Maybe you mean "associated value", instead of property?

yuzhe-han replied:
Good call out. I've changed this variable to be JS compliant. It's a implementation detail. I've added a appendix section at the bottom to explain the usage of this variable.

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
````diff
@@ -1,48 +1,70 @@
+ ```
+The second issue is that component creators can’t change the slotting behavior based on condition. A component that wants to display a subset of slottable based on a property can’t easily be done via declarative slotting.
+
+## Imperative Slotting API
````
This section contains no code examples, and jumps straight into a discussion of the spec mechanisms and internal ways in which something will be implemented. See https://github.com/w3ctag/w3ctag.github.io/blob/master/explainers.md for a bit more on how to write a good explainer, in particular by including example code showing the API solving problems---perhaps the two problems you've introduced in the beginning.

I would:

- Move the discussion of internal concepts and implementation details to an appendix near the end.
- Add some basic examples (not as long or detailed as the ones you have below) showing how this API solves the problem.

yuzhe-han replied:
Got it. Example addedd and moved the internal concepts to appendix at the bottom.

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
````diff
@@ -1,48 +1,70 @@
 
-See also the Example 3 later.
+```js
+const sr = attachShadow({ mode: ..., (optional) slotAssignment: 'manual' });
````
This isn't valid JavaScript syntax, so it's kind of confusing. I suggest:

```suggestion
const sr = attachShadow({ mode: 'open', slotAssignment: 'manual' });
```

yuzhe-han replied:
Update

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
```diff
@@ -54,36 +76,55 @@ partial interface HTMLSlotElement {
 
-Note: The detail is explained later, however, it would be worth noting that *manually-assigned-nodes* is not used as
-[assigned nodes] as is. You can think that `slot.assign(sequence<Node> nodes)` tell the engine *candidate nodes* from where [assigned nodes] are constructed.
+Note: The detail is explained later, however, it would be worth noting that `manually-assigned-nodes` is not used as
```
Where is "later"? Can you link to it?

yuzhe-han replied:
Removed.

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
````diff
@@ -54,36 +76,55 @@ partial interface HTMLSlotElement {
 
 
 ``` webidl
 ShadowRootInit {
````
It would be good to rewrite this closer to valid Web IDL.

yuzhe-han replied:
Updated with WebIDL

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
````diff
@@ -93,35 +134,33 @@ Other steps don't need to be updated from the standard's perspective, I think, t
-
 ```
 
 ``` javascript
````
The "== means ArrayEquals" is very confusing here since it means is isn't JavaScript code. It would be good to replace this all with `assert_array_equals`.

yuzhe-han replied:
Updated.

domenic reviewed (commented):
Exciting to see this being worked on! Some suggestions on how to make it better.

bicknellr commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
```diff
@@ -54,36 +76,55 @@ partial interface HTMLSlotElement {
+    
+    4. [Update Step] If parent is a shadow host, its shadow's `slotAssignment` is not "manual" and node is a slotable, then assign a slot for node.
+
+    5. [New Step] If parent's root is a shadow root, shadow's `slotAssignment` is "manual", and parent is a slot, throw exception.
```
This prevents slots in shadow roots with manual distribution from having fallback content. Is that intended?

yuzhe-han replied:
bicknellr@,

Yes, it's intended, but not sure if the spec wording is correct. Or if it needs to throw an exception vs no-op.

To me, fallback content makes it confusing when working in slot assignment manual mode. Slot assignment should be absolute and final in manual mode.  See comment from TPAC: https://github.com/whatwg/html/issues/3534#issuecomment-537802687

In that case, when developers call slot.assign([]), the slot should be cleared and have no content. But if we allow failback content, then content can still appear inside the slot even thou developer specifically want to clear it out. 

This is my reasoning for update the spec. Let me know if you think I'm overlooking something.

thanks

yuzhe-han replied:
bicknellr@,

After some considering and discussing it with masonfreed@, we decided it's better to preserve the fallback content behavior. So for a shadow root, regardless of whether slotAssignment is manual or auto, the fallback behavior is identical. 

I've since removed this section in the explainer doc. Thanks again for your review.

bicknellr replied:
Thanks for taking another look at this. I think that's the best option given that it makes the behavior the same in both cases and that slots' fallback content is very likely controlled by the same entity that would assign nodes.

yuzhe-han committed:
Update proposals/Imperative-Shadow-DOM-Distribution-API.md

Co-Authored-By: Domenic Denicola <d@domenic.me>

yuzhe-han committed:
Update proposals/Imperative-Shadow-DOM-Distribution-API.md

Co-Authored-By: Domenic Denicola <d@domenic.me>

yuzhe-han committed:
Update proposals/Imperative-Shadow-DOM-Distribution-API.md

Co-Authored-By: Domenic Denicola <d@domenic.me>

yuzhe-han committed:
Update Imperative-Shadow-DOM-Distribution-API.md

yuzhe-han committed:
Update Imperative-Shadow-DOM-Distribution-API.md

yuzhe-han committed:
Update Imperative-Shadow-DOM-Distribution-API.md

yuzhe-han committed:
Adds security and privacy section.

rniwa commented:
Looks like we never got to discuss this topic at this week's F2F.

@yuzhe-han: would you be interested in presenting / championing this proposal in the next virtual F2F if we ever have such a thing?

yuzhe-han commented:
> Looks like we never got to discuss this topic at this week's F2F.
> 
> @yuzhe-han: would you be interested in presenting / championing this proposal in the next virtual F2F if we ever have such a thing?

@rniwa: Sure. Would love a chance to present and meet you all.

yuzhe-han commented:
@domenic,

Hopefully, I've addressed all your issues. Can you please take another look?

Thanks.

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
```diff
@@ -54,210 +101,261 @@ partial interface HTMLSlotElement {
+dictionary ShadowRootInit {
+  required ShadowRootMode mode;
+  boolean delegatesFocus = false;
+  SlotAssignmentMode slotAssignment;   // optional. (defaults to 'auto')
```
```suggestion
  SlotAssignmentMode slotAssignment = "auto";
```

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
````diff
@@ -1,48 +1,95 @@
+ 
+#### Case 2: Slot based on condition. 
+```html
+<custom-tab show-panel=”2”>
````
```suggestion
<custom-tab show-panel="2">
```

domenic commented on proposals/Imperative-Shadow-DOM-Distribution-API.md:
```diff
@@ -1,48 +1,95 @@
+        div.innerHTML = template;
+        shadowRoot.appendChild(div);
+    }
+    connectedCallback() {
```
You might want to add a comment explaining that a production example would also use attributeChangeCallback to watch for changes to the `show-panel` attribute and re-run this code.

You could also just update the example to do that, but I'm not sure how much complexity it would add; perhaps it would be distracting.

yuzhe-han replied:
I've updated my example to include attributeChangedCallback.
Thanks.

domenic reviewed (approved):
LGTM with a couple nits; thank you!

I'd like to give you write access to this repo, so that you can merge on your own after fixing those nits, but I'm not sure who has permissions to do that. Maybe @plehegar can help.

yuzhe-han committed:
Update proposals/Imperative-Shadow-DOM-Distribution-API.md

Co-Authored-By: Domenic Denicola <d@domenic.me>

yuzhe-han committed:
Update proposals/Imperative-Shadow-DOM-Distribution-API.md

Co-Authored-By: Domenic Denicola <d@domenic.me>

yuzhe-han committed:
Update Imperative-Shadow-DOM-Distribution-API.md

Updated example usage to include attributeChangedCallback
